### DISCUSSION TITLE
Sepes vs PyTreeClass vs Equinox vs NNX

### DISCUSSION THREAD

**Discussion by adam-hartshorne** [ASEM000/sepes#19] [General]
I am just wondering how does Sepes library lie with pytreeclass and how all these compare with something like equinox? 

It seems like there are now many JAX / PyTree libraries (flax now has a new experiment nnx library for this) which give the user the ability to write more pythonic classes for JAX and under the hood they handle the PyTree. I am just trying to work out in my head why use one over another, and why you now have two libraries of your own that appear to do very similar things.

**Comment by ASEM000** [comment]
Thanks for your questions and sorry for late reply,

I marked pytreeclass stable for now, with no API change in the future. For sepes, It is continuation of pytreeclass with more room for experimentation.

IMO, One of the reasons why there is many jax libraries out there is because the jax nature of adopting many functional styles/assumption while being used in python. Each library has a bit different POV of how things should work along the line of being fully python from one end to being full functional to the other end. 

For instance NNX which is lead by Cristian and the flax team lies on the fully pythonic end. For context see this [PR](https://github.com/google/jax/issues/7919). By pythonic end, I mean reference sharing and statefulness is done in python style.

Equinox Module on the other hand is mixing and matching some python/pytorch concept with some functional concepts. Specifically, It allows for python/pytorch like in initialization and model definition, but does not allow reference sharing/stateful updates inside the method in python/pytorch style.

My work is to achieve pytorch/python style in both intialization/method calls with a transformation. See also [Oryx.nn](https://github.com/jax-ml/oryx/tree/main/oryx/experimental/nn) for closely related work.

> to work out in my head why use one over another

One of the ways to separate these libraries from each other is by implementing something like [Adam](https://serket.readthedocs.io/notebooks/%5Bguides%5D%5Bother%5Doptimlib.html) to see how each implements the stateful update step in the call method.

Another point, is how each library deal with jax transformation.
- Flax/Equinox/NNX wraps jax transformation to make their modules work.
- My work here, does not do this. Instead I mask the tree and unmask before call. Motivated by my observation that #function transformation/pytree accepting functions >> #function calls. 

Generally speaking, wrapped transformation are great for users who do not want to deal with the errors from passing non-jaxtypes in and out of the jax transformation. But the wrapped transformation does some implicit bookkeeping that goes against the explicit nature of jax. Additionally, going the wrapped route implies that as a library dev you have to wrap every jax function that accepts pytrees/function that accepts pytrees  or users will have to rewrite your code once you are hit with a code path that has non-wrapped function (new stuff like shard_map) or libraries that applies transformation inside their implementation (e.g. jaxopt). Finally you have to keep up with jax to update/deprecate your wrapped function, this means higher maintainability cost.

Let me know if this helps.
